src/frontend/src/expr/utils.rs

```diff
@@ -99,24 +99,72 @@ impl ExprRewriter for BooleanConstantFolding {
                 (rhs, lhs)
             }
         };
-        if contains_bool_constant {
-            match func_type {
-                Type::Not => {
-                    let constant_value = inputs.pop().unwrap();
-                    if let Some(v) = try_get_bool_constant(&constant_value) {
-                        return ExprImpl::literal_bool(!v);
-                    }
+        match func_type {
+            // unary functions
+            Type::Not => {
+                let input = inputs.first().unwrap();
+                if let Some(v) = try_get_bool_constant(input) {
+                    return ExprImpl::literal_bool(!v);
+                }
+            }
+            Type::IsFalse => {
+                let input = inputs.first().unwrap();
+                if input.is_null() {
+                    return ExprImpl::literal_bool(false);
                 }
-                Type::And => {
-                    let (constant_lhs, rhs) = prepare_binary_function_inputs(inputs);
-                    return boolean_constant_fold_and(constant_lhs, rhs);
+                if let Some(v) = try_get_bool_constant(input) {
+                    return ExprImpl::literal_bool(!v);
+                }
+            }
+            Type::IsTrue => {
+                let input = inputs.first().unwrap();
+                if input.is_null() {
+                    return ExprImpl::literal_bool(false);
                 }
-                Type::Or => {
-                    let (constant_lhs, rhs) = prepare_binary_function_inputs(inputs);
-                    return boolean_constant_fold_or(constant_lhs, rhs);
+                if let Some(v) = try_get_bool_constant(input) {
+                    return ExprImpl::literal_bool(v);
+                }
+            }
+            Type::IsNull => {
+                let input = inputs.first().unwrap();
+                if input.is_null() {
+                    return ExprImpl::literal_bool(true);
+                }
+            }
+            Type::IsNotTrue => {
+                let input = inputs.first().unwrap();
+                if input.is_null() {
+                    return ExprImpl::literal_bool(true);
                 }
-                _ => {}
+                if let Some(v) = try_get_bool_constant(input) {
+                    return ExprImpl::literal_bool(!v);
+                }
+            }
+            Type::IsNotFalse => {
+                let input = inputs.first().unwrap();
+                if input.is_null() {
+                    return ExprImpl::literal_bool(true);
+                }
+                if let Some(v) = try_get_bool_constant(input) {
+                    return ExprImpl::literal_bool(v);
+                }
+            }
+            Type::IsNotNull => {
+                let input = inputs.first().unwrap();
+                if try_get_bool_constant(input).is_some() {
+                    return ExprImpl::literal_bool(true);
+                }
```

Comment:
```suggestion
                if let v = try_get_bool_constant(input) {
                    return ExprImpl::literal_bool(v.is_some());
                }
```

Actually IsNotNull optimization can be applied on any constant folding, not just bool. I think eventually the code should be:

```rust
if let v = fold_constant(input) { // fn fold_constant(input) -> DatumRef
  return ExprImpl::literal_bool(v.is_some());
}
```

Reply:
> Actually IsNotNull optimization can be applied on any folded value, not just bool.

That makes sense. It should work on any non-null constant. I will fix it.

Reply:
Done in 5278ef0844e2e7822e9d4907daa4b927895e0d42.

`inputs` have been folded before(https://github.com/singularity-data/risingwave/pull/2265/files#diff-ae37135915c851ac6f44591a162bd326715ebbc24f279723c795826ddcae3232R85 FYI) so I only check whether `input` is null or not.